engine/classes/Elgg/UpgradeService.php

```diff
@@ -14,155 +14,168 @@
  */
 class UpgradeService {
 
-	/**
-	 * Global Elgg configuration
-	 *
-	 * @var \stdClass
-	 */
-	private $CONFIG;
+	private $events;
+
+	private $translator;
+
+	private $datalist;
+
+	private $logger;
+
+	private $db;
+
+	private $upgrade;
+
+	private $batch_run_time_in_secs;
 
 	/**
 	 * Constructor
 	 */
 	public function __construct() {
-		global $CONFIG;
-		$this->CONFIG = $CONFIG;
+		$this->events = _elgg_services()->events;
+		$this->translator = _elgg_services()->translator;
+		$this->datalist = _elgg_services()->datalist;
+		$this->logger = _elgg_services()->logger;
+		$this->db = _elgg_services()->db;
+		$this->session = _elgg_services()->session;
+
+		// TODO Make configurable
+		$batch_run_time_in_secs = 2;
 	}
 
 	/**
-	 * Run the upgrade process
+	 * Prepare the system to run an upgrade
 	 *
-	 * @return array
+	 * @param Upgrade $upgrade The upgrade object
+	 * @param int     $offset  Current offset of the items to upgrade
+	 * @return array $result Array containing possible errors
 	 */
-	public function run() {
+	public function prepareUpgrade(\Elgg\Upgrades\Upgrade $upgrade, $offset = null) {
+		$this->events->trigger('upgrade', 'system', array(
```

Comment:
Note to self: This doesn't belong here. It's not supposed to be run on every request of every individual upgrade.